### PR TITLE
[BugFix] fix prefixcache performance and accuracy on ascend

### PR DESCRIPTION
**Pull request by khalil2ji3mp6**
## Motivation

We have fixed some performance and accuracy problems in the MLA model on Ascend.

## Modifications

For performance problems:

1. We moved **device_indices** to the CPU in advance in *cache_controller.py*.
2. In *memory_pool_host.py*, we added a new variable **NEED_HOST_REGISTER** and **self.host_register** to indicate whether **cudaHostRegister** is required for pinning memory. If it is not needed, tensors are allocated directly with **pin_memory=True**.

For accuracy problems:

- We integrated the **npu_ring_mla** operator to perform segmented computation for the attention module.

## Accuracy Tests

We added a new test case, *test_ascend_hicache_mla.py*, to ensure the accuracy of the MLA model. The testing results are consistent with those obtained when the feature is disabled, indicating that the accuracy is normal.
Test case results:
image

## Benchmarking and Profiling

We tested the performance of the DeepSeek v3 model by comparing disabling the radix cache versus enabling L1 and L2 caches. The dataset used was 3.5K GSM8K with a 50% reuse rate (half of the requests are repeated), which ultimately **reduced TTFT by 39.5%**. Below are the specific execution and benchmark commands.

First run results:
```python
python3 -m sglang.launch_server \
    --model-path /xxx/dsv3 \
    --host xxx \
    --port 8000 \
    --trust-remote-code \
    --tp-size 16 \
    --mem-fraction-static 0.8 \
    --attention-backend ascend \
    --quantization w8a8_int8 \
    --device npu \
    --log-level debug \
    --disable-overlap-schedule \
    --max-running-requests 8 \
    --context-length 3800 \
    --chunked-prefill-size 57344 \
    --max-prefill-tokens 30400 \
    --disable-radix-cache
```
```python
python3 -m sglang.bench_serving \
    --dataset-path /xxx/GSM8K-in3584-bs100*2.jsonl \
    --dataset-name gsm8k \
    --backend sglang \
    --host xxx \
    --port 8000 \
    --max-concurrency 8 \
    --random-output-len 1 \
    --random-input-len 3584 \
    --num-prompts 200
```
image

Second run results:
```python
python3 -m sglang.launch_server \
    --model-path /xxx/dsv3 \
    --host xxx \
    --port 8000 \
    --trust-remote-code \
    --tp-size 16 \
    --mem-fraction-static 0.8 \
    --attention-backend ascend \
    --quantization w8a8_int8 \
    --device npu \
    --log-level debug \
    --disable-overlap-schedule \
    --max-running-requests 8 \
    --context-length 3800 \
    --chunked-prefill-size 57344 \
    --max-prefill-tokens 30400 \
    --enable-hierarchical-cache \
    --hicache-ratio 5 \
    --hicache-write-policy write_through
```
```python
python3 -m sglang.bench_serving \
    --dataset-path /xxx/GSM8K-in3584-bs100*2.jsonl \
    --dataset-name gsm8k \
    --backend sglang \
    --host xxx \
    --port 8000 \
    --max-concurrency 8 \
    --random-output-len 1 \
    --random-input-len 3584 \
    --num-prompts 200
```
image

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.ai/developer_guide/contribution_guide.html#code-style-guidance).
- [x] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/layers/attention/ascend_backend.py [hunk 1]
```diff
@@ -275,6 +287,24 @@ def init_forward_metadata(self, forward_batch: ForwardBatch):
 
         if forward_batch.forward_mode.is_target_verify():
             self.forward_metadata.seq_lens_cpu_int += self.speculative_num_draft_tokens
+
+        if (
+            forward_batch.forward_mode.is_extend()
+            and sum(forward_batch.extend_prefix_lens_cpu) > 0
+        ):
+            self.forward_metadata.prefix_lens = (
+                forward_batch.extend_prefix_lens.int().to("cpu")
```
Comment (khalil2ji3mp6) [inline python/sglang/srt/layers/attention/ascend_backend.py]: double check if exist extend_prefix_lens_cpu

File: python/sglang/srt/layers/attention/ascend_backend.py [hunk 2]
```diff
@@ -275,6 +287,24 @@ def init_forward_metadata(self, forward_batch: ForwardBatch):
...
+            )
+            seq_prefix_lens = self.forward_metadata.prefix_lens.tolist()
+            flatten_tlbs = torch.empty(0, dtype=torch.int32)
+            for req_idx, seq_len in zip(
+                forward_batch.req_pool_indices.tolist(), seq_prefix_lens
+            ):
+                indices = forward_batch.req_to_token_pool.req_to_token[req_idx]
+                tlbs = indices[:seq_len][:: self.page_size] // self.page_size
+                flatten_tlbs = torch.cat((flatten_tlbs, torch.flatten(tlbs).to("cpu")))
```
Comment (khalil2ji3mp6) [inline python/sglang/srt/layers/attention/ascend_backend.py]: change name

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py [hunk 3]
```diff
@@ -275,6 +287,24 @@ def init_forward_metadata(self, forward_batch: ForwardBatch):
 
         if forward_batch.forward_mode.is_target_verify():
             self.forward_metadata.seq_lens_cpu_int += self.speculative_num_draft_tokens
+
+        if (
```
Comment (iforgetmyname) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py:297]: check if only mla models need these

File: python/sglang/srt/layers/attention/ascend_backend.py [hunk 4]
```diff
@@ -69,6 +73,12 @@ def __init__(self, model_runner: ModelRunner, device, use_fia):
         mtp_mask_len = 2048
         self.mtp_mask = self.generate_mask_flag(mtp_mask_len).to(self.device)
 
+        # Initialize RingMla mask
+        ringmla_mask_len = 512
```
Comment (iforgetmyname) [inline python/sglang/srt/layers/attention/ascend_backend.py]: add if `mla` here

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py [hunk 5]
```diff
@@ -586,15 +616,99 @@ def forward_extend(
                         enable_gqa=use_gqa,
                         causal=causal,
                     )
+        elif sum(forward_batch.extend_prefix_lens_cpu) > 0:
```
Comment (iforgetmyname) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py:634]: check if this feature supports mtp

File: python/sglang/srt/mem_cache/memory_pool_host.py [hunk 6]
```diff
@@ -41,7 +41,7 @@
 
 logger = logging.getLogger(__name__)
 
-SUPPORT_PIN_MEMORY = not _is_npu
+NEED_HOST_REGISTER = not _is_npu
```
Comment (iforgetmyname) [inline python/sglang/srt/mem_cache/memory_pool_host.py]: use DefaultDict

pin_mem = defaultdict(
lamdba: host_register,
{"npu": torch_pin_memory}
)

pin_mem["npu"] = torch.empty(pin_memory=True)
pin_mem["cuda"] = torch.cuda.cudart().cudaHostRegister(

Comment (iforgetmyname) [reply python/sglang/srt/mem_cache/memory_pool_host.py]: register lamda function as value for defaultdict

File: test/srt/ascend/test_ascend_hicache_mla.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
+import unittest
```
Comment (iforgetmyname) [inline test/srt/ascend/test_ascend_hicache_mla.py:1]: cc @cherryblo

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py (hunk 3 above)
Comment (khalil2ji3mp6) [reply python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py:297]: Confirmed that only MLA models need this. Added the self.use_mla check accordingly.

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py (hunk 5 above)
Comment (khalil2ji3mp6) [reply python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py:634]: This feature can be used together with MTP. Since the KV cache in the MTP stage is relatively small, enabling prefix cache is not necessary for now.

File: python/sglang/srt/layers/attention/ascend_backend.py (hunk 1 above)
Comment (khalil2ji3mp6) [reply python/sglang/srt/layers/attention/ascend_backend.py]: extend_prefix_lens_cpu is a List[int], but what we need is a CPU-side tensor. This conversion is unavoidable at the moment.

File: python/sglang/srt/layers/attention/ascend_backend.py (hunk 2 above)
Comment (khalil2ji3mp6) [reply python/sglang/srt/layers/attention/ascend_backend.py]: We have renamed tlbs to req_prefix_block_tables, and flatten_tlbs to flatten_prefix_block_tables, to better indicate the prefix block tables for a single request and the flattened prefix block tables for all requests.

File: python/sglang/srt/layers/attention/ascend_backend.py (hunk 4 above)
Comment (khalil2ji3mp6) [reply python/sglang/srt/layers/attention/ascend_backend.py]: The use_mla check has already been added.

File: python/sglang/srt/mem_cache/memory_pool_host.py (hunk 6 above)
Comment (khalil2ji3mp6) [reply python/sglang/srt/mem_cache/memory_pool_host.py]: I’ve changed the host-side memory allocation into ALLOC_MEMORY_FUNCS, which returns different allocation functions based on the device type. For example:
```python
ALLOC_MEMORY_FUNCS = defaultdict(
    lambda: alloc_default,
    {
        "npu": alloc_npu,
    },
)
```

File: python/sglang/srt/mem_cache/memory_pool_host.py [hunk 8]
```diff
@@ -53,6 +52,28 @@ def wrapper(self, *args, **kwargs):
     return wrapper
 
 
+def alloc_manual_pinned(dims, dtype, device, pin_memory):
```
Comment (iforgetmyname) [inline python/sglang/srt/mem_cache/memory_pool_host.py]: `alloc_with_host_register`

File: python/sglang/srt/mem_cache/memory_pool_host.py [hunk 9]
```diff
@@ -53,6 +52,28 @@ def wrapper(self, *args, **kwargs):
...
+    buffer = torch.empty(dims, dtype=dtype, device=device)
+    if pin_memory:
+        torch.cuda.cudart().cudaHostRegister(
+            buffer.data_ptr(), buffer.numel() * buffer.element_size(), 0
+        )
+    return buffer
+
+
+def alloc_native_pinned(dims, dtype, device, pin_memory):
```
Comment (iforgetmyname) [inline python/sglang/srt/mem_cache/memory_pool_host.py]: `alloc_with_pin_memory`

File: python/sglang/srt/mem_cache/memory_pool_host.py (hunk 8 above)
Comment (iforgetmyname) [reply python/sglang/srt/mem_cache/memory_pool_host.py]: p.s. add typing hints for these functions

File: python/sglang/srt/mem_cache/memory_pool_host.py (hunk 9 above)
Comment (khalil2ji3mp6) [reply python/sglang/srt/mem_cache/memory_pool_host.py]: Same as above.

File: python/sglang/srt/mem_cache/memory_pool_host.py (hunk 8 above)
Comment (khalil2ji3mp6) [reply python/sglang/srt/mem_cache/memory_pool_host.py]: I have already renamed the functions based on your suggestions and added typing hints. And I think that dims is not suitable for typing hints, so I decided to drop the parameter validation for it.

**Review by iforgetmyname — APPROVED** [review]
